### PR TITLE
Implemented group re-arrangement.

### PR DESCRIPTION
**Pull request by theduckylittle**
New MOVE_GROUP action will change the place of a group
with-in the layers list.

### CODE REVIEW

**Comment by theduckylittle** [comment]
Looking at #735, this may or may not be needed as it handles the 'special case' of groups more seamlessly.

File: src/reducers/map.js [hunk 1]
```diff
@@ -170,6 +170,89 @@ function orderLayer(state, action) {
...
+  // get the real index of the next spot for the group,
+  // if the placeAt index is mid group then place_at should
+  // be the index of the FIRST member of that group.
+  let place_start = place_at;
+  const place_group = getGroup(state.layers[place_at]);
+
+  if (place_group) {
+    // find the first index of the place group
+    let new_place = -1;
+    for (let i = 0, ii = n_layers; i < ii && new_place < 0; i++) {
```
Comment (bartvde) [inline src/reducers/map.js]: redefine of i

File: src/reducers/map.js [hunk 2]
```diff
@@ -170,6 +170,89 @@ function orderLayer(state, action) {
...
+      if (getGroup(state.layers[i]) === place_group) {
+        new_place = i;
```
Comment (bartvde) [inline src/reducers/map.js]: personally I'd prefer a break here instead of the check in the for loop

File: src/reducers/map.js [hunk 3]
```diff
@@ -170,6 +170,89 @@ function orderLayer(state, action) {
...
+  }
+
+
+  // build a new array for the layers.
+  let new_layers = [];
+
+  // have the group layers ready to concat.
+  const group_layers = state.layers.slice(group_start, group_end + 1);
+
+  for (let i = 0, ii = n_layers; i < ii; i++) {
```
Comment (bartvde) [inline src/reducers/map.js]: another redefine of i and ii

File: src/reducers/map.js [hunk 4]
```diff
@@ -561,6 +644,9 @@ function setZoom(state, action) {
   return Object.assign({}, state, {zoom});
 }
 
+/** Move an entire group of layers in the stack
+ *
```
Comment (bartvde) [inline src/reducers/map.js]: is this comment in the correct place?

File: __tests__/reducers/map-groups.test.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    const layer_ids = [];
+    for (let i = 0, ii = layers.length; i < ii; i++) {
+      layer_ids.push(layers[i].id);
+    }
+
+    expect(layer_ids).toEqual(expectedOrder);
+  }
+
+  function createDefaultLayers() {
+    createLayer('A');
```
[the remark this answers is not in the document]
Comment (theduckylittle) [reply __tests__/reducers/map-groups.test.js]: I didn't do that to allow for different tests for different combinations of layers and groups.

[the remark this answers is not in the document]
Comment (bartvde) [reply __tests__/reducers/map-groups.test.js]: ok, so with future tests in my mind

File: src/reducers/map.js [hunk 6]
```diff
@@ -170,6 +170,89 @@ function orderLayer(state, action) {
...
+  let group_end = null;
+  for (let i = 0, ii = n_layers; i < ii; i++) {
+    const group = getGroup(state.layers[i]);
+    if (group === action.group) {
+      if (group_start === null || i < group_start) {
+        group_start = i;
+      }
+      if (group_end === null || i > group_end) {
+        group_end = i;
+      }
```
[the remark this answers is not in the document]
Comment (theduckylittle) [reply src/reducers/map.js]: https://jsperf.com/math-min-max-vs-ternary-vs-if -- On my machine, using if()s are performing ~80% faster than calling out to Math.min.

**Pull request merged**